lib/dor/models/publishable.rb

```diff
@@ -10,6 +10,58 @@ module Publishable
+    # @return [String] the computed thumb filename, with the druid prefix and a slash in front of it, e.g. oo000oo0001/filenamewith space.jp2
+    def thumb
+       return if contentMetadata.nil?
+       cm = contentMetadata.ng_xml
+       mime_type_finder = "@mimetype='image/jp2' or @mimeType='image/jp2'" # allow the mimetype attribute to be lower or camelcase when searching to make it more robust
+       thumb_image=nil
+
+       # these are the finders we will use to search for a thumb resource in contentMetadata, they will be searched in the order provided, stopping when one is reached
+       thumb_xpath_finders = [
+           {image_type: 'local', finder: "/contentMetadata/resource[@type='thumb' and @thumb='yes']/file[#{mime_type_finder}]"},      # first find a file of mimetype jp2 explicitly marked as a thumb in the resource type and with a thumb=yes attribute
+           {image_type: 'external', finder: "/contentMetadata/resource[@type='thumb' and @thumb='yes']/externalFile[#{mime_type_finder}]"}, # same thing for external files
+           {image_type: 'local', finder: "/contentMetadata/resource[(@type='page' or @type='image') and @thumb='yes']/file[#{mime_type_finder}]"},# next find any image or page resource types with the thumb=yes attribute of mimetype jp2
+           {image_type: 'external', finder: "/contentMetadata/resource[(@type='page' or @type='image') and @thumb='yes']/externalFile[#{mime_type_finder}]"},# same thing for external file
+           {image_type: 'local', finder: "/contentMetadata/resource[@type='thumb']/file[#{mime_type_finder}]"}, # next find a file of mimetype jp2 or resource type=thumb but no marked with the thumb directive
+           {image_type: 'external', finder: "/contentMetadata/resource[@type='thumb']/externalFile[#{mime_type_finder}]"}, # same thing for external file
+           {image_type: 'local', finder: "/contentMetadata/resource[@type='page' or @type='image']/file[#{mime_type_finder}]"}, # finally find the first very first page or image resource of mimetype jp2
+           {image_type: 'external', finder: "/contentMetadata/resource[@type='page' or @type='image']/externalFile[#{mime_type_finder}]"} # same thing for external file
+         ]
+              
+       thumb_xpath_finders.each do |search_path|
+         thumb_files = cm.xpath(search_path[:finder]) # look for a thumb  
+         if thumb_files.size > 0   # if we find one, return the filename based on whether it is a local file or external file
+           if search_path[:image_type] == 'local'
+             thumb_image="#{remove_druid_prefix}/#{thumb_files[0]['id']}"
+           else
+             thumb_image="#{remove_druid_prefix(thumb_files[0]['objectId'])}/#{thumb_files[0]['fileId']}"
+           end
+           break  # break out of the loop so we stop searching
+         end
+       end 
+              
+       thumb_image
```

Comment:
agreed that caching in memory might cause odd problems if contentMetadata changes, but using a local variable in the method so avoid multiple calls to thumb is totally reasonable and i'll do that